kernel/comps/virtio/src/queue.rs

```diff
@@ -187,22 +151,182 @@ impl VirtQueue {
         if inputs.is_empty() && outputs.is_empty() {
             return Err(QueueError::InvalidArgs);
         }
-        if inputs.len() + outputs.len() + self.num_used as usize > self.queue_size as usize {
+        if inputs.len() + outputs.len() > self.desc_table.num_free_desc() {
+            return Err(QueueError::BufferTooSmall);
+        }
+
+        let desc_head = self.desc_table.add_dma_buf(inputs, outputs);
+        self.aval_ring.push_desc(desc_head);
+        Ok(desc_head)
+    }
+
+    /// Adds buffers to the virtqueue, return a token. **This function will be removed in the future.**
+    ///
+    /// Ref: linux virtio_ring.c virtqueue_add
+    pub fn add_buf(&mut self, inputs: &[&[u8]], outputs: &[&mut [u8]]) -> Result<u16, QueueError> {
+        // FIXME: use `DmaSteam` for inputs and outputs. Now because the upper device driver lacks the
+        // ability to safely construct DmaStream from slice, slice is still used here.
+        // pub fn add(
+        //     &mut self,
+        //     inputs: &[&DmaStream],
+        //     outputs: &[&mut DmaStream],
+        // ) -> Result<u16, QueueError> {
+
+        if inputs.is_empty() && outputs.is_empty() {
+            return Err(QueueError::InvalidArgs);
+        }
+        if inputs.len() + outputs.len() > self.desc_table.num_free_desc() {
             return Err(QueueError::BufferTooSmall);
         }
 
-        // allocate descriptors from free list
+        let desc_head = self.desc_table.add_buf(inputs, outputs);
+        self.aval_ring.push_desc(desc_head);
+        Ok(desc_head)
+    }
+
+    /// Returns whether there is an used element that can pop.
+    pub fn can_pop(&self) -> bool {
+        self.used_ring.can_pop()
+    }
+
+    /// Returns the number of free descriptors.
+    pub fn num_free_desc(&self) -> usize {
```

Comment:
I believe that Descriptors is an internal concept of the virt queue. Assuming this is true, it should not be exposed through a public method.